In [2]:
import pandas as pd

## Reading files

In [3]:
wh2018 = pd.read_csv('2018_v2.csv')
wh2019 = pd.read_csv('2019_v2.csv')

Adding extra column to identify year of population

In [4]:
wh2018['year'] = 2018
wh2019['year'] = 2019

Union of three files

In [5]:
wh = pd.concat([wh2018,wh2019])

In [6]:
wh.columns

Index(['Overall rank', 'Country or region', 'Score', 'GDP per capita',
       'Social support', 'Healthy life expectancy',
       'Freedom to make life choices', 'Generosity',
       'Perceptions of corruption', 'year'],
      dtype='object')

## Ejercicios

1. Partiendo de Wh, realice un analisis de valores nulos. que porcentaje de valores nulos se encuentran en cada columna por anio. presente el analisis en un dataframe

Ejemplo:

In [61]:
pd.DataFrame({
    "year": [2017,2017]
    , "column" : ["Happiness.Rank","Whisker.low"]
    , "percentage.of.nans" : [0.2,0.2]
})

year          column  percentage.of.nans
0  2017  Happiness.Rank                 0.2
1  2017     Whisker.low                 0.2

### resolucion ejercicio uno

In [ ]:
##Visualizando los datos Nulos

In [37]:
wh.isnull()


Overall rank  Country or region  Score  GDP per capita  Social support  \
0           False              False  False           False           False   
1           False              False  False           False           False   
2           False              False  False           False           False   
3           False              False  False           False            True   
4           False              False  False           False           False   
..            ...                ...    ...             ...             ...   
151         False              False  False           False           False   
152         False              False  False           False           False   
153         False              False  False           False           False   
154         False              False  False           False           False   
155         False              False  False           False           False   

     Healthy life expectancy  Freedom to make life choices  Generosity  \
0                      False                         False       False   
1                      False                         False       False   
2                      False                         False       False   
3                      False                         False        True   
4                      False                         False       False   
..                       ...                           ...         ...   
151                    False                         False       False   
152                    False                         False       False   
153                    False                         False       False   
154                    False                         False       False   
155                    False                         False       False   

     Perceptions of corruption   year  
0                        False  False  
1                        False  False  
2                        False  False  
3                        False  False  
4                        False  False  
..                         ...    ...  
151                      False  False  
152                      False  False  
153                      False  False  
154                      False  False  
155                      False  False  

[312 rows x 10 columns]

In [ ]:
## Cantidad de datos nulos por columna 

In [39]:
wh.isnull().sum()

Overall rank                     0
Country or region                0
Score                           19
GDP per capita                  12
Social support                  16
Healthy life expectancy         26
Freedom to make life choices    23
Generosity                      21
Perceptions of corruption       13
year                             0
dtype: int64

In [ ]:
##Porcentaje de datos nulos por columna

In [40]:
(wh.isnull().sum()/len(wh))*100

Overall rank                    0.000000
Country or region               0.000000
Score                           6.089744
GDP per capita                  3.846154
Social support                  5.128205
Healthy life expectancy         8.333333
Freedom to make life choices    7.371795
Generosity                      6.730769
Perceptions of corruption       4.166667
year                            0.000000
dtype: float64

In [44]:
wh.groupby(['year']).size().reset_index(name='numero de registros')

year  numero de registros
0  2018                  156
1  2019                  156

2. partiendo de Wh, por año muestre un resumen de todas las columnas.

nota: aunque solo tenemos tres años, que pasaria si en nuestro dataframe tenemos 20 años? resuelva su ejercicio lo mas dinamico que se pueda

### resolucion ejercicio dos

In [47]:
n=2019
wh[wh['year']==n]

Overall rank         Country or region  Score  GDP per capita  \
0               1                   Finland  7.769           1.340   
1               2                   Denmark  7.600           1.383   
2               3                    Norway  7.554           1.488   
3               4                   Iceland  7.494           1.380   
4               5               Netherlands  7.488           1.396   
..            ...                       ...    ...             ...   
151           152                    Rwanda  3.334           0.359   
152           153                  Tanzania  3.231           0.476   
153           154               Afghanistan  3.203           0.350   
154           155  Central African Republic  3.083           0.026   
155           156               South Sudan  2.853           0.306   

     Social support  Healthy life expectancy  Freedom to make life choices  \
0             1.587                      NaN                         0.596   
1             1.573                    0.996                         0.592   
2             1.582                    1.028                         0.603   
3             1.624                    1.026                         0.591   
4             1.522                    0.999                         0.557   
..              ...                      ...                           ...   
151           0.711                    0.614                         0.555   
152           0.885                    0.499                         0.417   
153           0.517                    0.361                         0.000   
154           0.000                    0.105                         0.225   
155           0.575                    0.295                         0.010   

     Generosity  Perceptions of corruption  year  
0         0.153                      0.393  2019  
1         0.252                      0.410  2019  
2         0.271                      0.341  2019  
3         0.354                      0.118  2019  
4         0.322                      0.298  2019  
..          ...                        ...   ...  
151       0.217                      0.411  2019  
152       0.276                      0.147  2019  
153       0.158                      0.025  2019  
154       0.235                      0.035  2019  
155       0.202                      0.091  2019  

[156 rows x 10 columns]

In [48]:
n=2018
wh[wh['year']==n]


Overall rank         Country or region  Score  GDP per capita  \
0               1                   Finland  7.632           1.305   
1               2                    Norway  7.594           1.456   
2               3                   Denmark  7.555           1.351   
3               4                   Iceland  7.495           1.343   
4               5               Switzerland  7.487           1.420   
..            ...                       ...    ...             ...   
151           152                     Yemen  3.355           0.442   
152           153                  Tanzania  3.303           0.455   
153           154               South Sudan  3.254           0.337   
154           155  Central African Republic  3.083           0.024   
155           156                   Burundi  2.905           0.091   

     Social support  Healthy life expectancy  Freedom to make life choices  \
0             1.592                    0.874                         0.681   
1             1.582                    0.861                         0.686   
2             1.590                    0.868                         0.683   
3               NaN                    0.914                         0.677   
4             1.549                    0.927                         0.660   
..              ...                      ...                           ...   
151           1.073                    0.343                         0.244   
152             NaN                    0.381                         0.481   
153             NaN                    0.177                         0.112   
154           0.000                    0.010                         0.305   
155           0.627                    0.145                         0.065   

     Generosity  Perceptions of corruption  year  
0         0.202                      0.393  2018  
1         0.286                      0.340  2018  
2         0.284                      0.408  2018  
3           NaN                      0.138  2018  
4         0.256                      0.357  2018  
..          ...                        ...   ...  
151       0.083                        NaN  2018  
152       0.270                      0.097  2018  
153       0.224                      0.106  2018  
154       0.218                      0.038  2018  
155       0.149                      0.076  2018  

[156 rows x 10 columns]

3. partiendo de Wh, remplace los valores NaN de las columnas por su mean, el mean debe de ser calculado por año para ser mas exacto

### resolucion ejercicio tres

In [58]:
scor=wh['Score']
scor.fillna(scor.mean(),inplace=True)




In [59]:
scor

0      7.632
1      7.594
2      7.555
3      7.495
4      7.487
       ...  
151    3.334
152    3.231
153    3.203
154    3.083
155    2.853
Name: Score, Length: 312, dtype: float64

NameError: name 'capita' is not defined

4. Partiendo de Wh limpio, para cada metrica por pais, muestra la diferencia del ultimo valor encontrado contra el primero. 

Ejemplo:

In [62]:
wh.loc[ wh['Country or region'] == 'Switzerland' , : ]

Overall rank Country or region  Score  GDP per capita  Social support  \
4             5       Switzerland  7.487           1.420           1.549   
5             6       Switzerland  7.480           1.452           1.526   

   Healthy life expectancy  Freedom to make life choices  Generosity  \
4                    0.927                         0.660       0.256   
5                    1.052                         0.572       0.263   

   Perceptions of corruption  year  
4                      0.357  2018  
5                      0.343  2019

Como se puede observar para Switzerland en Score la diferencia seria 7.480 - 7.487

Claramente en su solucion la columna año se debera de eliminar

### resolucion ejercicio cuatro